### PR TITLE
User data preprocessor

### PR DESCRIPTION
**Pull request by weilixu**
Preprocessor + validator

### CODE REVIEW

**Comment by weilixu** [comment]
@dmaddoxwhite @lymereJ I separate this ticket into two functions
1. Preprocessing function: this function preprocess the format to clean up the data - for example, trim the string or lower case
2. Validation function: valid the user data to test any errors in the data for example, the fraction of total lighting types shall equal to 1.0.

**Comment by lymereJ** [comment]
Looks good to me! I just have two comments:
- Instead of `assert()` I would suggest that we use `OpenStudio.logFree(OpenStudio::Error, '', '')` to be consistent with the rest of the code base and so errors would show up in the log which might be used by third party tools/software.
- My personal preference would be to make `user_data_validation()` more modular, i.e. call a different method for each verification. It might be easier for future expansion and review. I'm fine as it is though.

File: lib/openstudio-standards/standards/ashrae_90_1_prm/ashrae_90_1_prm.rb [hunk 1]
```diff
@@ -87,4 +90,72 @@ def load_userdata_to_standards_database(json_path)
...
+        unless lpd_row['num_std_ltg_types'].to_i == 0
+          num_ltg_type = lpd_row['num_std_ltg_types'].to_i
+          total_ltg_percent = 0.0
+          std_ltg_index = 0
+          while std_ltg_index < num_ltg_type
+            frac_key = format('std_ltg_type_frac%02d', (std_ltg_index + 1))
+            total_ltg_percent += lpd_row[frac_key].to_f
+            std_ltg_index += 1
+          end
+          raise `The fraction of user defined lighting types in Space/SpaceType: #{lpd_row['name']} does not add up to 1.0. The calculated fraction is #{total_ltg_percent}.` if (total_ltg_percent - 1.0).abs > 0.01
```
[the remark this answers is not in the document]
Comment (lymereJ) [reply lib/openstudio-standards/standards/ashrae_90_1_prm/ashrae_90_1_prm.rb:122]: It for sure shows up in the stdout but I'm not sure about the lof. I think we should probably follow existing patterns like this one where both `raise` and `OpenStudio.logFree()` are used:
https://github.com/NREL/openstudio-standards/blob/abae01541a336d08f736eb00fad9fba24fea4f23/lib/openstudio-standards/prototypes/common/prototype_metaprogramming.rb#L81-L82

**Comment by lymereJ** [comment]
> @weilixu - I think that the OpenStudio error will stop the process and log the error, that's why I suggested it.

I guess this is not true, it doesn't stop the process.

File: lib/openstudio-standards/standards/ashrae_90_1_prm/ashrae_90_1_prm.rb (hunk 1 above)
[the remark this answers is not in the document]
Comment (weilixu) [reply lib/openstudio-standards/standards/ashrae_90_1_prm/ashrae_90_1_prm.rb:122]: This is what I am thinking - use OpenStudio log to log the message as well as raise the exception to stop the program.
I need to try out the logFree function to see how we can add the User Data Error as the label of the error msg.

**Comment by weilixu** [comment]
> > @weilixu - I think that the OpenStudio error will stop the process and log the error, that's why I suggested it.
> 
> I guess this is not true, it doesn't stop the process.

Thanks for checking. I will add both to the code

**Comment by weilixu** [comment]
@dmaddoxwhite @lymereJ 

I was thinking another implementation.
Create a flag: `user_data_validate_outcome = true` and Loop through all user data. In the loop:
2. OpenStudio.log if any errors, and `user_data_validate_outcome = false`
3. In the end, check `user_data_validate_outcome`
4. If `user_data_validation_outcome == false`, raise exception with a message: User data invalid, check log file to see detail error messages.
5. If `user_data_validation_outcome == true`, then continue processing.

This will change the execution logic slightly but gives an overview of all the issues detected in the user data and stop before the baseline automation is called.

**Comment by weilixu** [comment]
The function is updated.
The validation will run through all user data and log errors if found.
In the end, if one or more errors found in user data, the validation would fail and raise exception.

**Review by dmaddoxwhite — APPROVED** [review]

**Pull request merged**